### DISCUSSION TITLE
How does OpenSumi  extension  achieve data/style isolation

### DISCUSSION THREAD

**Discussion by scarqin** [opensumi/core#547] [Q&A]
I want to know how the ui-based extension implements data/style isolation. If it runs in the rendering process, won't the extension error affect the main application?

**Comment by Aaaaash** [comment]
All ui-based extension run in ShadowRoot.

![image](https://user-images.githubusercontent.com/17701805/156540459-29791d2a-e9d5-4755-b498-902cc4176233.png)
